### DISCUSSION TITLE
How to update data after a short delay in a Story

### DISCUSSION THREAD

**Discussion by dcrackel** [storybookjs/storybook#26558] [Help] [answered]
### Summary

I have a component that has animations that fire when certain data conditions are meet.  I have two mock data files, after a short delay I would like to switch one for the other to show the animation effect.

```
export const DynamicDataUpdate: Story = {
  render: args => ({
    component: LeaderboardColumnComponent,
    props: args,
    onInit: () => {
      setTimeout(() => {
        args.leaderboardData = [...MockLeaderboardDataLongTermTwo];
      }, 2000); 
    }
  }),
  args: {
    leaderboardData: MockLeaderboardDataLongTerm
  }
};
```

I've tried a bunch of different ways to swap the data source after a delay, without luck. I'm sure others have done this. 

What's the trick?

Thank you for your help.

**Comment by Marklb** [comment, marked as answer]
To me, this sounds like something that you don't want to do by changing args, because they only reset back to the initial when refreshing the page, I think. You can update args by using the `useArgs` hook, exported by `@storybook/preview-api`, but I think the places that use that are also the places that would get run again if you update args. So triggering it from a timeout would be an infinite loop, without some condition that you could maybe come up with, but I think it would be more complicated than necessary for this.

I don't know what framework you are using, to provide an example in your framework, but would you be fine to wrap your story in a decorator that handles the logic for delayed updating or using a feature of your framework? Someone else may have a better idea, but I will give a potential example for Angular, since that is the framework I most often use.

```ts
// This is just a pipe that I quickly made, which emits an initial value then after a delay it emits a different value.
@Pipe({
  name: 'delay',
  standalone: true,
  pure: true,
})
class DelayPipe implements PipeTransform {
  transform(value: string, delayMs: number, valueAfterDelay: string): Observable<string> {
    return of(undefined).pipe(
      switchMap(() => timer(delayMs).pipe(map(() => valueAfterDelay))),
      startWith(value),
    )
  }
}

// This story uses the pipe and provides the "initial" and "after" values in args, so they could be changed independently in controls.
export const Delay: StoryObj<ButtonComponent & { valueAfterDelay: string }> = {
  render: (args) => ({
    props: args,
    template: `<storybook-button ${argsToTemplate(args, { exclude: ['label'] })} [label]="label | delay: 1000:valueAfterDelay | async"></storybook-button>`,
    moduleMetadata: {
      imports: [ DelayPipe ]
    },
  }),
  args: {
    size: 'small',
    label: 'Button',
    valueAfterDelay: 'After',
  },
};
```

Another option could be to wrap your story in a decorator that maybe adds a button or something to trigger the change then in the stories `play` function you could trigger it after a delay.
